core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -224,23 +231,33 @@ public synchronized void schedulingComplete()
             return;
         }
 
-        if (isFlushing()) {
-            stateMachine.transitionToFlushing();
+        try {
+            allTasksLock.readLock().lock();
+            if (isFlushing()) {
+                stateMachine.transitionToFlushing();
+            }
+            if (isFinished()) {
+                stateMachine.transitionToFinished();
+            }
         }
-        if (finishedTasks.containsAll(allTasks)) {
-            stateMachine.transitionToFinished();
+        finally {
+            allTasksLock.readLock().unlock();
         }
 
         for (PlanNodeId partitionedSource : stage.getFragment().getPartitionedSources()) {
             schedulingComplete(partitionedSource);
         }
     }
 
-    private synchronized boolean isFlushing()
+    private boolean isFlushing()
     {
         // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
-        return !flushingTasks.isEmpty()
-                && allTasks.stream().allMatch(taskId -> finishedTasks.contains(taskId) || flushingTasks.contains(taskId));
+        return runningTasksCounter.get() == -1;
```

Comment:
this is super disturbing :) This commit increases complexity a lot (I think above my PR review confidence). I am not convinced we want that.

Reply:
Well, the most safe, readable and easy to reason about approach is just to use `synchronized`, I agree. But, as a side effect we can observe high contention, what we was actually did. The code was rewritten to use `lock-free` data structures and atomic counters. The decision to move the machine state from `ANY_STATE` to `FLUSHING/FINISHED` is basing on the one specific counter (`flushing / finished`) to make it easier to reason about.

We have a high lock contention what limits scaling capabilities. In some moment we need to do something with that. I would say that it is expected that lock-free code is more complex that the "lock-full" one.